dlib/matrix/fft_size.h

```diff
@@ -0,0 +1,215 @@
+            DLIB_ASSERT(index < _size, "index " << index << " out of range [0," << _size << ")");
+            return _dims[index];
+        }
+        /*!
+            requires
+                - size() > 0
+                - index < size()
+            ensures
+                - returns a const reference to the element at position index
+        !*/
+        
+        const_reference back() const
+        {
+            DLIB_ASSERT(_size > 0, "object is empty");
+            return _dims[_size-1];
+        }
+        /*!
+            requires
+                - size() > 0
+            ensures
+                - returns a const reference to (*this)[size()-1]
+        !*/
+        
+        long dimprod() const
+        {
+            DLIB_ASSERT(_size > 0, "object is empty");
+            return std::accumulate(begin(), end(), 1, std::multiplies<long>());
+        }
+        /*!
+            requires
+                - size() > 0
+            ensures
+                - returns the product of all dimensions, i.e. the total number
+                  of elements.
+        !*/
+        
+        bool operator==(const fft_size& other) const
+        {
+            return this->_size == other._size && std::equal(begin(), end(), other.begin());
+        }
+        /*!
+            ensures
+                - returns true if two fft_size objects have same size and same elements.
+        !*/
+        
+        bool is_valid() const
+        {
+            return _size > 0 && std::find_if(begin(), end(), [](long dim) {return dim <= 0;}) == end();
+        }
+        /*!
+            ensures
+                - returns true if:
+                    - size() > 0
+                    - (*this) contains strictly positive elements
+        !*/
+        
+        /*global helpers*/
+        friend inline dlib::uint32 hash(
+            const fft_size& item,
+            dlib::uint32 seed = 0)
+        {
+            seed = dlib::hash((dlib::uint64)item.size(), seed);
+            seed = std::accumulate(item.begin(), item.end(), seed, [](dlib::uint32 seed, long next) {
+                return dlib::hash((dlib::uint64)next, seed);
+            });
+            return seed;
+        }
+        /*!
+            ensures
+                - returns a 32bit hash of the data stored in item.
+        !*/
+        
+        friend inline fft_size squeeze_ones(const fft_size& size)
+        {
+            fft_size newsize;
+            if (size.dimprod() == 1)
+            {
+                newsize = {1};
+            }
+            else
+            {
+                newsize = size;
+                const auto newend = std::remove(newsize.begin(), newsize.end(), 1);
+                const long nremoved = std::distance(newend, newsize.end());
+                newsize._size -= nremoved;
+            }
+            return newsize;
+        }
+        /*!
+            ensures
+                - returns a copy of size with all but 1 elements equal to 1 removed
+        !*/
```

Comment:
The dlib style is to put the comment above the code, so like this
```
void foo()
/*!
   comments
!*/
{
...
}
```

Also, `returns a copy of size with all but 1 elements equal to 1 removed` is ungrammatical.   How about something like: removes dimensions with values equal to 1, yielding a new size with the same dimprod() but fewer dimensions.